### PR TITLE
Add source_output_file_map_aspect and DataStore symlink logic

### PR DESCRIPTION
**Pull request by thiagohmcruz**
* Collect mapping information (from source code to respective `.o` file under `bazel-out`) to existing aspect implemented in `BazelExtensions/xcode_configuration_provider.bzl`
* Adds `XCHammerAssets/bazel_build_service_setup.sh` to setup many things necessary to setup the build service changes implemented here: https://github.com/jerrymarino/xcbuildkit/pull/49
* Adds `bazel_build_service_config` attr to allow one to configure the build service
* Load build service configurations in Xcode so the new pre build script introduced here can update the `xcbuildkit` config file on the fly
* Bump `Tulsi` to pick up latest fixes there ([this](https://github.com/bazel-ios/tulsi/pull/7), [this](https://github.com/bazel-ios/tulsi/pull/8))

### CODE REVIEW

File: BazelExtensions/xcodeproject.bzl [hunk 1]
```diff
@@ -185,6 +195,14 @@ def _install_xcode_project_impl(ctx):
         "(rm -f $SRCROOT/external && ln -sf $PWD/../../external $SRCROOT/external)",
         'echo "' + output_proj + '" > ' + ctx.outputs.out.path,
     ]
+
+    if enable_indexing and relative_index_store_path:
+        command.extend([
+            "DD=$(xcodebuild -project " + output_proj + " -showBuildSettings 2>&1 | grep -e \"^\\s*BUILT_PRODUCTS_DIR = \" | cut -d'=' -f2 | xargs | sed -e 's/\\/Build.*//g')",
```
Comment (jerrymarino) [inline BazelExtensions/xcodeproject.bzl]: Is it not possible to set this via `xcconfig` in Xcode?

File: BazelExtensions/xcodeproject.bzl [hunk 2]
```diff
@@ -185,6 +195,14 @@ def _install_xcode_project_impl(ctx):
...
+            'mv $DD/Index/DataStore $DD/Index/DataStore.default || true',
+            "mkdir -p '$SRCROOT{}'".format(relative_index_store_path),
+            "ln -s $(echo $SRCROOT){} $DD/Index/DataStore".format(relative_index_store_path),
```
Comment (jerrymarino) [inline BazelExtensions/xcodeproject.bzl]: re - this one: https://github.com/bazel-ios/xchammer/pull/33/files#r998629066

File: BazelExtensions/source_output_file_map_aspect.bzl [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    return [
+        OutputGroupInfo(
+            source_output_file_map = depset([source_output_file_map] + transitive_jsons),
+        ),
+        SourceOutputFileMapInfo(
+            mapping = mapping
+        ),
+    ]
+
+source_output_file_map_aspect = aspect(
```
Comment (jerrymarino) [inline BazelExtensions/source_output_file_map_aspect.bzl]: A couple suggestion:
- can we align this with the existing aspect to prevent another traversal
- do you want to tack this onto the `xcodeproj` rule `deps` attr - so that these files are generated at _project_ generation time or not?

File: BazelExtensions/source_output_file_map_aspect.bzl [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    # List of source files to be mapped to output files
+    if hasattr(ctx.rule.attr, "srcs"):
+        objc_srcs = [
+            f
+            for source_file in ctx.rule.attr.srcs
+            for f in source_file.files.to_list()
+            # Handling objc only for now
+            if f.path.endswith(".m")
+            # TODO: handle swift
```
Comment (jerrymarino) [inline BazelExtensions/source_output_file_map_aspect.bzl]: ➕ you'll need to handle how records work with `strategy=SwiftCompile=WORKER` - which will implicate assumptions on the rules here or there

File: BazelExtensions/xcodeproject.bzl (hunk 1 above)
Comment (thiagohmcruz) [reply BazelExtensions/xcodeproject.bzl]: Done

File: BazelExtensions/xcodeproject.bzl (hunk 2 above)
Comment (thiagohmcruz) [reply BazelExtensions/xcodeproject.bzl]: Done

File: BazelExtensions/source_output_file_map_aspect.bzl (hunk 4 above)
Comment (thiagohmcruz) [reply BazelExtensions/source_output_file_map_aspect.bzl]: Yep, will work on this next and put up a separate PR. Don't want to increase the scope here.

File: BazelExtensions/source_output_file_map_aspect.bzl (hunk 3 above)
Comment (thiagohmcruz) [reply BazelExtensions/source_output_file_map_aspect.bzl]: >can we align this with the existing aspect to prevent another traversal

Done

>do you want to tack this onto the xcodeproj rule deps attr - so that these files are generated at project generation time or not?

That would be helpful, to have the aspect output groups ready on Xcode launch. I played with this for a bit but hit issues due to not being able to pre-generate in a transition mappings containing the correct `bazel-out` paths. Essentially the `ST-hash` was different on "aspect + transition" vs "XCHammer build". I believe this is a bug but haven't spent more time digging, since everything works once at least a build in Xcode happens I think we can punt this for now.

TL;DR from my investigation: building with cmd line options `--apple_platform_type=ios --cpu=ios_sim_arm64` is **not** the same as using a transition that transitions to the *exactly same* flags. I believe this is related to this: https://github.com/bazelbuild/bazel/issues/12171

File: BazelExtensions/xcode_configuration_provider.bzl [hunk 5]
```diff
@@ -154,8 +154,79 @@ def _install_action(ctx, infos, itarget):
...
+    objc_objects = []
+
+    # List of source files to be mapped to output files
+    if hasattr(ctx.rule.attr, "srcs"):
+        objc_srcs = [
+            f
+            for source_file in ctx.rule.attr.srcs
+            for f in source_file.files.to_list()
+            # Handling objc only for now
+            if f.path.endswith(".m")
```
Comment (jerrymarino) [inline BazelExtensions/xcode_configuration_provider.bzl]: Missing other lang types ? `.c` `.cpp` `.mm` `.cxx` ?

File: BazelExtensions/xchammerconfig.bzl [hunk 6]
```diff
@@ -56,13 +56,22 @@ def target_config(
 def project_config(
         paths,  # [String]?
         build_bazel_platform_options = None,  # [String: [String]]?
+        bazel_build_service_config = None, # [String: XCHammerTargetConfig]?
```
Comment (jerrymarino) [inline BazelExtensions/xchammerconfig.bzl:59]: The data type here is `BazelBuildServiceConfig`?

File: BazelExtensions/xcode_configuration_provider.bzl [hunk 7]
```diff
@@ -154,8 +154,79 @@ def _install_action(ctx, infos, itarget):
...
+    transitive_jsons = []
+    for dep in deps:
+        # Collect mappings from deps
+        if SourceOutputFileMapInfo in dep:
+            for k, v in dep[SourceOutputFileMapInfo].mapping.items():
+                mapping[k] = v
+        # Collect generated JSON files from deps
+        if OutputGroupInfo in dep:
+            if hasattr(dep[OutputGroupInfo], "source_output_file_map"):
+                transitive_jsons.extend(dep[OutputGroupInfo].source_output_file_map.to_list())
```
Comment (jerrymarino) [inline BazelExtensions/xcode_configuration_provider.bzl]: Do you need to call `to_list()` here - or can you append a depset and pass it to the `depset` consturctor `depset(.., transitive=transtive_jsons)` where `transitive_jsons is an array of `depset`

File: XCHammerAssets/bazel_build_service_setup.sh [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+if [[ ! -d $WORKSPACE_ROOT ]]; then
+    echo "[ERROR] Failed to setup build service. Workspace not found at $WORKSPACE_ROOT"
+    exit 1
+fi
+
+INDEX_DATA_STORE=$WORKSPACE_ROOT/Index/DataStore
+INDEX_DATA_STORE_BACKUP=$WORKSPACE_ROOT/Index/DataStore.default
+
+undoDataStoreSetupIfNecessary () {
```
Comment (jerrymarino) [inline XCHammerAssets/bazel_build_service_setup.sh]: From a structure perspective do you think it is better to run this script from the build service? It seems like we can encapsulate such an implementation detail of indexing there / not a concern a project generator necessarily needs to deal with

Comment (jerrymarino) [reply XCHammerAssets/bazel_build_service_setup.sh]: Mainly, the the bulk of purpose of the project generator would be to write the config for a given project - and the build service would do any of this symlinking

File: XCHammerAssets/bazel_build_service_setup.sh [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+# If there's an existing DataStore create a backup before symlink-ing
+if [[ -d $INDEX_DATA_STORE ]]; then
+    if [[ -d $INDEX_DATA_STORE_BACKUP ]]; then
+        rm -fr $INDEX_DATA_STORE_BACKUP
+    fi
+    mv $INDEX_DATA_STORE $INDEX_DATA_STORE_BACKUP
+fi
+
+ln -s $BUILD_SERVICE_INDEX_STORE_PATH $INDEX_DATA_STORE
```
Comment (jerrymarino) [inline XCHammerAssets/bazel_build_service_setup.sh]: I'd suggest adding a comment about why it does this

**Review by jerrymarino — APPROVED** [review]
Overall LGTM besides minor edit to type / inline comment. I'd suggest to consider looking at the encapsulation of this logic into the other program but possibly as a followup since you have it working

File: XCHammerAssets/bazel_build_service_setup.sh (hunk 8 above)
Comment (thiagohmcruz) [reply XCHammerAssets/bazel_build_service_setup.sh]: Yep, this makes a lot of sense. Moved the DataStore setup logic to `xcbuildkit`. See latest commit: https://github.com/jerrymarino/xcbuildkit/pull/49/commits/1eb4ed9f2026b35cece333308ba9623acede90f3

File: XCHammerAssets/bazel_build_service_setup.sh (hunk 9 above)
Comment (thiagohmcruz) [reply XCHammerAssets/bazel_build_service_setup.sh]: Moved the DataStore logic to `xcbuildkit` and added a comment about this, see latest commit: https://github.com/jerrymarino/xcbuildkit/pull/49/commits/1eb4ed9f2026b35cece333308ba9623acede90f3

File: BazelExtensions/xchammerconfig.bzl (hunk 6 above)
Comment (thiagohmcruz) [reply BazelExtensions/xchammerconfig.bzl:59]: Yes, check that type in `Sources/XCHammer/XCHammerConfig.swift`

File: BazelExtensions/xcode_configuration_provider.bzl (hunk 5 above)
Comment (thiagohmcruz) [reply BazelExtensions/xcode_configuration_provider.bzl]: Good catch, will add the other types here using `rules_ios` as a reference

File: BazelExtensions/xcode_configuration_provider.bzl (hunk 7 above)
Comment (thiagohmcruz) [reply BazelExtensions/xcode_configuration_provider.bzl]: good catch, will fix that!
